sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Main.java

```diff
@@ -74,30 +80,92 @@
  * <p>See <a href="http://datalab.cs.pdx.edu/niagaraST/NEXMark/">
  * http://datalab.cs.pdx.edu/niagaraST/NEXMark/</a>
  */
-public class Main<OptionT extends NexmarkOptions> {
+public class Main {
+
+  private static class Run implements Runnable {
+    final NexmarkLauncher<NexmarkOptions> nexmarkLauncher;
+    final NexmarkConfiguration configuration;
+    final BlockingQueue results;
+    NexmarkPerf perf = null;
+    Exception exception = null;
+
+    private Run(String[] args, NexmarkConfiguration configuration, BlockingQueue results) {
+      NexmarkOptions options = PipelineOptionsFactory.fromArgs(args).as(NexmarkOptions.class);
+      this.nexmarkLauncher = new NexmarkLauncher<>(options);
+      this.configuration = configuration;
+      this.results = results;
```

Comment:
Can you avoid using global blocking state by using `ExecutorService.submit(Callable)` override? https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html#submit(java.util.concurrent.Callable)

With `Callable` you are allowed to return the value which you then extract from the `Future`

Reply:
Actually this is a good idea. By doing so you can also avoid that count, because you will try to get result from a list of Future one by one.